src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
         {
             get
             {
+                if (validValuesGeneratorType != null)
+                {
+                    ValidValuesCacheEntry validValuesCacheEntry;
+                    if (validValuesCache.TryGetValue(validValuesGeneratorType, out validValuesCacheEntry))
+                    {
+                        var currentTime = DateTime.Now;
+                        if (DateTime.Compare(validValuesCacheEntry.cacheEntryCreationTime.AddSeconds(CacheExpiration), currentTime) < 0)
+                        {
+                            // Update expired Entry.
+                            validValuesCacheEntry.validValues = validValuesCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+                            validValuesCacheEntry.cacheEntryCreationTime = currentTime;
+                            validValuesCache[validValuesGeneratorType] = validValuesCacheEntry;
+                        }
+                    }
+                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```

Comment:
Sorry, I thought I had replied.

I now think `CacheExpiration` should **not** be an argument to the attribute because, as you point out, different uses might specify conflicting values.  Instead, the generator should own the cache expiration policy, and the attribute should not cache the results from the generator.

I still like the idea of providing an abstract base class that implements a default caching policy - it would be optional for a generator to derive from it.